File: android/java/app/src/main/java/com/daily/partyline/JoinFragment.java
package com.daily.partyline;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;

import android.text.Editable;
import android.text.TextWatcher;
import android.text.method.LinkMovementMethod;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;

import com.daily.partyline.databinding.FragmentJoinBinding;

import static android.content.Context.INPUT_METHOD_SERVICE;

public class JoinFragment extends Fragment {

    FragmentJoinBinding binding;

    public JoinFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        binding = FragmentJoinBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    // Post view initialization logic
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {

        binding.learnMore.setMovementMethod(LinkMovementMethod.getInstance());

        NavController navController = Navigation.findNavController(view);
        JoinFragmentDirections.ActionJoinFragmentToRoomFragment toRoom = JoinFragmentDirections.actionJoinFragmentToRoomFragment();

        binding.roomUrl.getEditText().addTextChangedListener(new TextWatcher() {

            @Override
            public void afterTextChanged(Editable s) {}

            @Override
            public void beforeTextChanged(CharSequence s, int start,
                                          int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start,
                                      int before, int count) {
                if (s.length() != 0) {
                    binding.joinButton.setText("Join room");
                }
                else {
                    binding.joinButton.setText("Create and join room");
                }
            }
        });

        binding.joinButton.setOnClickListener((sender) -> {
            String userName = getUserName();
            String roomUrl = getRoomCode();

            if (!userName.matches("") && roomUrl.matches("")) {
                toRoom.setUserName(userName);
                toRoom.setRoomUrl(null);
            }
            else if (!userName.matches("")) {
                toRoom.setUserName(userName);
                toRoom.setRoomUrl(roomUrl);
            }

            dismissKeyboard();
            navController.navigate(toRoom);
        });
    }

    //region utils
    private String getUserName() {
        String firstName = binding.firstName.getEditText().getText().toString().replace(" ","");
        String lastName = binding.lastName.getEditText().getText().toString().replace(" ","");;
        return (firstName.equals("") ? "Test" : firstName) + (lastName.equals("") ? "" : " " + lastName);
    }

    private String getRoomCode() {
        return binding.roomUrl.getEditText().getText().toString().trim();
    }

    private void dismissKeyboard() {
        requireActivity().runOnUiThread(() -> {
            InputMethodManager imm = (InputMethodManager) requireActivity().getSystemService(INPUT_METHOD_SERVICE);
            if(imm.isAcceptingText()) { // verify if the soft keyboard is open
                imm.hideSoftInputFromWindow(requireActivity().getCurrentFocus().getWindowToken(), 0);
            }
        });
    }
    //endregion utils
}

File: android/kotlin/app/src/main/java/com/daily/partyline/JoinFragment.kt
package com.daily.partyline

import android.content.Context
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.text.method.LinkMovementMethod
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import androidx.fragment.app.Fragment
import androidx.navigation.Navigation
import com.daily.partyline.databinding.FragmentJoinBinding

class JoinFragment : Fragment() {
    lateinit var binding : FragmentJoinBinding
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        binding = FragmentJoinBinding.inflate(inflater, container, false)
        return binding.root
    }

    // Post view initialization logic
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        binding.learnMore.movementMethod = LinkMovementMethod.getInstance()
        val navController = Navigation.findNavController(view)
        val toRoom: JoinFragmentDirections.ActionJoinFragmentToRoomFragment = JoinFragmentDirections.actionJoinFragmentToRoomFragment()
        binding.roomUrl.editText?.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable?) {}
            override fun beforeTextChanged(s: CharSequence?, start: Int,
                                           count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence?, start: Int,
                                       before: Int, count: Int) {
                if (s?.length != 0) {
                    binding.joinButton.text = "Join room"
                } else {
                    binding.joinButton.text = "Create and join room"
                }
            }
        })
        binding.joinButton.setOnClickListener { sender: View? ->
            val userName = getUserName()
            val roomUrl = getRoomCode()
            if (userName?.isEmpty() == false && roomUrl?.isEmpty() == true) {
                toRoom.setUserName(userName)
                toRoom.setRoomUrl(null)
                dismissKeyboard()
                navController.navigate(toRoom)
            } else if (userName?.isEmpty() == false && roomUrl?.isEmpty() == false) {
                toRoom.setUserName(userName)
                toRoom.setRoomUrl(roomUrl)
                dismissKeyboard()
                navController.navigate(toRoom)
            }
        }
    }

    //region utils
    private fun getUserName(): String? {
        val firstName = binding.firstName.editText?.getText().toString().replace(" ", "")
        val lastName = binding.lastName.editText?.getText().toString().replace(" ", "")
        return (if (firstName == "") "Test" else firstName) + if (lastName == "") "" else " $lastName"
    }

    private fun getRoomCode(): String? {
        return binding.roomUrl.editText?.getText().toString().trim()
    }

    private fun dismissKeyboard() {
        requireActivity().runOnUiThread {
            val imm = requireActivity().getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            if (imm.isAcceptingText) { // verify if the soft keyboard is open
                imm.hideSoftInputFromWindow(requireActivity().currentFocus.windowToken, 0)
            }
        }
    } //endregion utils
}